### DISCUSSION TITLE
Improve Pull Request and Release procedures

### DISCUSSION THREAD

**Discussion by fernandezc** [spectrochempy/spectrochempy#434] [Developer discussions]
**! WORK IN PROGRESS**

The release of a new version of SpectroChemPy is subject to some errors that cannot be changed after the release: e.g., the assignment of a DOI by zenodo is final! It is therefore important, following my various recent experiences, to improve the release process of a new version. 

I think that the first thing, even if it can seem tedious, is to control what is pushed on the master branch, as far as possible of course because some bugs are difficult to detect. 

# Prerequisites 
Here are some prerequisites that we could propose (Don't hesitate to add comments, and I will modify this list if necessary:
1. Compatibility (except major impossibility) with the latest version of Python - currently 3.10 - A test to pass is therefore required to be able to merge a PR. 
2. Compatibility with the version used by Google Colab (currently 3.7). This is important because, one goal would be to write part of the documentation in this form, in order to make it easier for new users to get started. 
3. The CHANGELOG.md should have been updated with an `Unreleased` section to be completed with the changes: NEW FEATURES, BUG FIXED, DEPRECATED, REMOVED, DEPENDENCE MODIFIED, among others.
For instances:

```
# What's new

## Unreleased

### NEW FEATURES
* enable custom widget in COLAB
* New worflow for publishing new releases

## Version 0.4.3
``` 
 A workflow during the release will introduce the version string automatically. Automatic generation of changelog items could be considered, but the discipline of writing commit messages needs to be improved.
For example follow the recommendation in our documentation:  [Commit your Code](https://www.spectrochempy.fr/latest/devguide/contributing.html#commit-your-code). 

As a reminder, prefix commits with:
* ENH or DEV: Enhancement, new feature
* FIX: Bug fixes
* DOC: Documentation additions/updates
* TEST: Test additions/updates
* BUILD: Build process/script updates
* PERF: Performance improvements
* MAINT: Code cleanup

(This list can be corrected, extended or reduced!)

4. To enforce these recommendations, approval is required to merge and resolve all discussions in a PR. This will apply to everyone of course :-)

# Automation of the release process

The workflow scripts are inspired by the blog of : [Thomas Eizinger](https://blog.eizinger.io/12274/using-github-actions-to-automate-gitflow-style-releases)

The procedure is the following (not following the more complex GitFlow process ):
* All changes have to be send to the upstream master branch through a Pull Request (PR).
* The master branch is updated (by rebasing)  after required tests pass and approval have been given.

## Draft a new release (workflow `draft_new_release.yml`)
* To make the final release, we automate the process by creating a `release/{version}` branch from the  master.
* Update the `CHANGELOG.md`, especially with the desired version number, update the `requirements.txt` and `environments.yml` if necessary, as well as the `.zenodo.json` and `CITATION.cff` files which need to be updated with the date and version number.
* Commit and push on the release branch.
* Then we open a PR on the master branch.
* The tests are executed and if all goes well we accept to proceed to the  **merge** (this is important to trigger the following publish_new_release workflow).

## Publish  a new release (workflow `publish_new_release.yml`)
* We then tag the merge commit with the right version number.
* We release the new version which also means publishing it on zenodo.org, anaconda.org and pypi.

**Comment by atravert** [comment]
This looks fine to me.
-  I have a question regarding the prefix commits. In some case cases a given commit or PR will be both FIX and ENH, what should be done in such case ? -- but I will (try to) stick strictly in this from now on :-)

**Reply by fernandezc** [reply]
I would say: if we want to put it in BUG FIXED in the changelogs -> FIX 
If it ha s to go in NEW FEATURE ->  ENH. 

But this is just a proposal, as I said, it can be adapted and modified. This is just to have a clear history. In many cases, I think a PR should be merged with REBASE and MERGE, because several types of changes may have been made to solve the test problem, and it is good to have this information in the history.

**Reply by fernandezc** [reply]
One thing about approvals. It may be required, but I feel like it's a bit much to have to go through all the PRs in detail. That's why I think we will not requiring it at this time.

**Reply by fernandezc** [reply]
Oh another thing, the last realease 0.4.3 was done automatically from the develop branch that I added in SpectroChempy repository. In the new processus I'm setting up, everyone should push their modifications in this branch. We will only push in the master branch using the automatic process to make releases. I will describe all this as well as possible in this discussion, when I will have finished.

**Reply by fernandezc** [reply]
In view of the new workflow, I did not set the approval requirement for the Develop branch, but only for the Master branch. It seems indeed reasonable that from now, a Release PR is discussed first. Sorry, I didn't follow until now 😄 .
